packages/roosterjs-editor-plugins/lib/CustomReplace/CustomReplace.ts

```diff
@@ -0,0 +1,157 @@
+import { Editor, EditorPlugin } from 'roosterjs-editor-core';
+import { PluginEvent, PluginEventType, PluginKeyPressEvent } from 'roosterjs-editor-types';
+
+type Replacement = {
+    sourceString: string,
+    replacementHTML: string,
+    matchSourceCaseSensitive: boolean
+}
+
+const defaultReplacements: Replacement[] = [
+    {
+        sourceString: ':)',
+        replacementHTML: '🙂',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ';)',
+        replacementHTML: '😉',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ':O',
+        replacementHTML: '😲',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ':o',
+        replacementHTML: '😯',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: '<3',
+        replacementHTML: '❤️',
+        matchSourceCaseSensitive: true,
+    },
+]
+
+/**
+ * Wrapper for CustomReplaceContentEditFeature that provides an API for updating the
+ * content edit feature
+ */
+export default class CustomReplacePlugin implements EditorPlugin {
+    private longestReplacementLength: number;
+    private editor: Editor;
+
+    /**
+     * Create instance of CustomReplace plugin
+     * @param features An optional feature set to determine which features the plugin should provide
+     */
+    constructor(private replacements: Replacement[] = defaultReplacements) {
+        this.longestReplacementLength = getLongestReplacementSourceLength(this.replacements);
+    }
+
+    updateReplacements(newReplacements: Replacement[]) {
+        this.replacements = newReplacements;
+        this.longestReplacementLength = getLongestReplacementSourceLength(this.replacements);
+    }
+
+    /**
+     * Get a friendly name of  this plugin
+     */
+    getName() {
+        return 'CustomReplace';
+    }
+
+    /**
+     * Initialize this plugin
+     * @param editor The editor instance
+     */
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose(): void {
+    }
+
+    public onPluginEvent(event: PluginEvent) {
+        if (this.editor.isInIME() || event.eventType != PluginEventType.KeyPress) {
+            return;
+        }
+
+        // This handler fires before the event's native behaviour -- defer until after it completes
+        setTimeout(() => this.handlePluginEvent(event), 0)
+    }
+
+    private handlePluginEvent(event: PluginKeyPressEvent) {
+        // Get the matching replacement
+        const range = document.getSelection().getRangeAt(0);
+        if (range == null) return;
+        const searcher = this.editor.getContentSearcherOfCursor();
```

Comment:
Suggest use cacheGetContentSearcher(), it will returns a cached version if any, of a new one otherwise.

Reply:
What I saw while debugging using cacheGetContentSearcher was:

1. I start with the text `the quick brown fo` in my editor
2. I press the `x` key
3. rooster wraps the event and passes it through plugins
4. This plugin schedules an event for after the event is processed.
5. another plugin calls `cacheGetContentSearcher` and attaches it to the event. The PositionContentSearcher saves the current position of the selection(immediately after the 'o')
6. the event handler stack resolves & the browser native behaviour kicks in
7. the async handler fires, inserts an 'x' after the 'o', and moves the selection to the right 1 character
8. the async handler calls `cacheGetContentSearcher`, which fetches the same PositionContentSearcher off the event's cache. That position is no longer correct, since it is still saved to where the 'o' was.

Reply:
Ok, that makes sense. I also saw this happen to me before when develop auto bullet.